### PR TITLE
Coqui AI and Tortoise TTS

### PR DESCRIPTION
**Pull request by da3dsoul**
As per #885, Coqui has Pros and Cons. It's already written, so PRing it. More soon (tm). When this is accepted, I'll try to write up some more info into the wiki....and figure out how to do that.

### CODE REVIEW

**Comment by da3dsoul** [comment]
Tortoise needs more testing by not me. I don't have the hardware to run it alongside any of the models I have. As per discussion in #885, this has 3 Tortoise implementations: official, fast, and MRQ

**Comment by da3dsoul** [comment]
I can't do anything about Mac support, since a lot of this stuff uses CUDA (nvidia GPU acceleration). I found the quality of Coqui varies quite a lot depending on which model you use, as it has several choices. The speed and minimum requirements, as well....

**Comment by Ph0rk0z** [comment]
Coqui is good enough.. it generates a ok-ish voice in a few seconds. Tortoise would need its own GPU.. it's not even about memory, it just takes a while.

**Comment by St33lMouse** [comment]
Coqui is about the same quality as Silero, which we already have. It's fine to support it, but I don't think it brings much to new to the table. Tortoise does. And it looks like Bark will too.

**Comment by da3dsoul** [comment]
Coqui actually does have things it brings to the table. It has contextual text normalization of numbers and much better  cadence and stressing of syllables in English. This leads to a much more natural sounding sample, even if the audio quality is about the same.
It also is faster than Tortoise, but supports custom voices, unlike Silero, even if it isn't as high quality as Tortoise. I think it's a actually a good in-between in terms of speed, spec requirements (GPU), and quality.

**Comment by Ph0rk0z** [comment]
Coqui can clone, silero can't. The cloning isn't that great but it's fast.

**Comment by G-force78** [comment]
Hi, many thanks for the work on this voice cloning is super interesting, I am trying to test tortoise mrq version in the 4bit colab on free tier but failing to load the extension, have added the extension flag but it is missing things, any ideas what to add would be much appreciated

```
Traceback (most recent call last):
  File "/content/text-generation-webui/modules/extensions.py", line 34, in load_extensions
    exec(f"import extensions.{name}.script")
  File "<string>", line 1, in <module>
  File "/content/text-generation-webui/extensions/tortoise_tts_mrq/script.py", line 11, in <module>
    from modules import chat, shared, tts_preprocessor
ImportError: cannot import name 'tts_preprocessor' from 'modules' (unknown location)
```

**Comment by da3dsoul** [comment]
It's in modules/tts_preprocessor.py

**Comment by da3dsoul** [comment]
Took a bit of a break. @oobabooga aside from merging and fixing the conflicts, what do you want done with this?
I think it's pretty much ready

**Comment by oobabooga** [comment]
@da3dsoul two questions:

1) Can you give me a quick rundown on the conclusions of your testing? Have you adopted any of these as your daily driver instead of silero?
2) Could it be possible to create a general "tts" extension with an output modifier and several submodules, one for each engine, like `extensions/tts/silero.py`, `extensions/tts/elevenlabs.py`, etc? Or is that too ambitious?

**Comment by da3dsoul** [comment]
1. Yes, my thoughts on it from before hold up. Coqui is generally faster and less resource intensive. Tortoise (standard) is most likely to get updates and improve. Tortoise Fast is actually fast enough for me to consider using it. Tortoise MRQ has the most features.
I have been running Tortoise Fast, as it's enough of an improvement over Coqui to take the speed hit imo.
2. I think it's possible, but should be part of a more specialized and generified project. You would have more of an extension pipeline.


#### Pipeline:
1. registration (tell the system if the extension is able to be set up and what the extension does. Good for first time users and UI)
2. UI setup (mostly as we have it. It should be lightweight loading of things like dropdowns and defaults)
4. pre-LLM (input modifiers)
5. model loading (this step should be separate and can be ran at startup or on-demand with universal model swapping)
6. LLM (we could theoretically generify the LLM systems and make them extensions)
7. Post-LLM (output modifiers, pre-tts, could allow tts_preprocessors as extensions)
8. TTS (takes the text as input and returns the audio file path. The audio player will be added by the core code)
9. Model unloading (like the loading step. Could just not be ran or ran after every query for model swapping)
10. Raw output modifiers (new step. For modifying the resulting html if desired)

Because of the registration step, each extension should have a separate file, register.py. In it, it should have methods for determining if the extension has prerequisites installed, as well as defining the next part.
Each extension would have metadata set in the register.py, such as a way to tell it the order to run in the pipeline (priority), a type (TTS, LLM, input/output processor, etc), a variable storing the state and if it errors (good for UI), and some basic UI info to display to the user, like a name, description, etc

**Comment by BuffMcBigHuge** [comment]
I was able to test out your plugins in an afternoon. Unfortunately the install procedures have been very tricky, as installing each plugin results in dependency conflicts with each other. I was able to get `coqui_tts` and `tortoise_tts_mrq` running, however tortoise refused to perform inference after the models loaded. No errors, just no audio. My suggestion would be to have an install similar to `bark_tts` as a separate repo, where the extension can be installed on it's own.

Thanks for sharing your work, looking forward to testing new TTS implementations.

**Comment by da3dsoul** [comment]
Did you use the install scripts? They have some commands that edit the requirements.txt to not cause conflicts with the base repo. I'm not sure about Coqui and Tortoise, as Coqui is installed via pip, making its requirements handled separately

**Comment by BuffMcBigHuge** [comment]
Yup I used the install script provided. 

The primary issue is that `coqui_tts` and it's single pip dependency 'TTS' has major conflicts with the tortoise installs. I had to manually run pip commands to solve these conflicts. I also had to manually download vocoder.pth and put it in /models/tortoise, as it wasn't downloading on it's own.

Just now, I was able to get `tortoise_tt_fast` working with some finagling. It's great!

**Comment by Ph0rk0z** [comment]
I can say, having used silero more regularly now in tavern, it is fast but very meh. On one GPU this was rather all slow.. but now that I have 2, I'm going to see if the situation improves when dedicating a card to TTS specifically. 

If you're going to have a TTS, might as well have a good one.

**Comment by ksylvan** [comment]
@BuffMcBigHuge can you resolve the conflicts here so I can checkout this PR and test it? Excited about using Coqui

**Comment by da3dsoul** [comment]
> @BuffMcBigHuge can you resolve the conflicts here so I can checkout this PR and test it? Excited about using Coqui

Generally, it's the person at the top that you would tag. Sure I can update it

File: .gitignore [hunk 1]
```diff
@@ -3,9 +3,10 @@ characters
 training/datasets
 extensions/silero_tts/outputs/*.wav
 extensions/elevenlabs_tts/outputs/*.wav
-extensions/sd_api_pictures/outputs/*
 extensions/tortoise_tts*/tortoise/**/*
```
Comment (Ph0rk0z) [inline .gitignore:6]: I added 
extensions/*/outputs/*.*
extensions/*/outputs/*

I wonder if that works correctly

Comment (da3dsoul) [reply .gitignore:6]: My reasoning for not doing that is that it'll cause weird conflicts with the "output_goes_here.txt" files

Comment (Ph0rk0z) [reply .gitignore:6]: ahh.. but do we really need those?

Comment (da3dsoul) [reply .gitignore:6]: No, but I didn't want to be the one to remove them haha

**Comment by da3dsoul** [comment]
Fix those conflicts finally. @ksylvan

**Comment by Urammar** [comment]
Hey sorry if im annoying here, im a bit of a github noob. This seems awesome and I want to try out these extensions. Can someone help me out? How on earth do I download this pull request on its own so I just have the files and can place them where they need to be? Very confused.

Thank you!

**Comment by Urammar** [comment]
> Hey sorry if im annoying here, im a bit of a github noob. This seems awesome and I want to try out these extensions. Can someone help me out? How on earth do I download this pull request on its own so I just have the files and can place them where they need to be? Very confused.
> 
> Thank you!

Okay I straight up asked GPT, learned git and pulled it. Time to break things! Thanks for the massive effort you guys, will let you know how it goes!

**Comment by Urammar** [comment]
image

So I ran the scrips in the extension directory (for coqui I think its just pip install -r requirements.txt in that directory?)

I'm guessing this means i dont have a model for this, tts_fast does a very similar thing, but theres no information on where to get one or where to put it once I have it? Little lost now, have to say, sorry all.

**Comment by da3dsoul** [comment]
I've never seen that one before. Coqui TTS has models downloaded by itself. I can maybe try to reproduce it today.

Tortoise Fast just needs the models copied from normal Tortoise. I'm not sure why they don't include them in the repo.

**Comment by Urammar** [comment]
image

Okay so... was trying to run test.py as... you know.. a test?

Seems like the issue is that tts cant be found, but if I type tts on its own it gives me the syntax for using it?

image

**Comment by da3dsoul** [comment]
I wonder if Coqui updated and moved something

**Comment by Urammar** [comment]
Also as far as Tortoise

image

I am really having no luck at all!

**Comment by da3dsoul** [comment]
Huh which model is that? It does tell you what to check. BigVGAN is missing. I've never heard of that, but that's what it says

**Comment by da3dsoul** [comment]
> I wonder if Coqui updated and moved something

Can confirm this is not the case. I installed fresh and it worked

**Comment by Urammar** [comment]
> Huh which model is that? It does tell you what to check. BigVGAN is missing. I've never heard of that, but that's what it says

Thats just what it says out of the box trying to load the addon in the webui, I have no idea what it is referencing, but its hardcoded in the vocoder.py?

`import torch
import torch.nn as nn
import torch.nn.functional as F

import json
from enum import Enum
from typing import Optional, Callable
from dataclasses import dataclass
try:
    from BigVGAN.models import BigVGAN as BVGModel
    from BigVGAN.env import AttrDict
except ImportError:
    raise ImportError(
        "BigVGAN not installed, can't use BigVGAN vocoder\n"
        "Please see the installation instructions on README."
    )

MAX_WAV_VALUE = 32768.0`

**Comment by Urammar** [comment]
I downloaded pytorch_model.bin from [here](https://huggingface.co/Gatozu35/tortoise-tts/tree/main) and threw it in the models folder, but that didnt change anything either.

**Comment by Urammar** [comment]
Literally all I did was click this and apply and restart interface

image

Shouldnt there be some kind of model loader box, dropdown, or something? Instead of just erroring out in the terminal?

**Comment by da3dsoul** [comment]
There should, yes, and that's why I'm confused

**Comment by Urammar** [comment]
Okay so ive done a complete reinstall of ooba, and its working much better now, it even grabbed the model by itself which is nice, and correctly installed those dependencies.

Im now getting the following error when I generate text, though, despite the extension actually loading in the ui and everything else seeming like its playing nice.

It does indeed generate text, but no audio.

**Comment by da3dsoul** [comment]
I'll see if I can make that happen and let you know. Can you show what extensions you have loaded?

**Comment by Urammar** [comment]
I created a brand new ooba specifically just for coqui also, and ive run that separately (not both at once) in their own little conda environments and that says this

Capture

-------

> I'll see if I can make that happen and let you know. Can you show what extensions you have loaded?

Yeah, so for Tortoise_tts_fast is literally just that and I think I had gallery checked also. 

Capture

--------

So to reiterate, these are brand new installs of the oobabooga webui with nothing in them except a small model for testing and the relevant extension

**Comment by Urammar** [comment]
Capture

update:
After rebooting the webui again trying to launch tortoisettsfast, im now getting totally different errors.

This seems *really* janky man, are you trying to merge this pull request? This is absolutely not ready for prime time. Have you only tested this on your own machine? Does this literally only work in your environment?

**Comment by da3dsoul** [comment]
Ok I'll take a look

**Comment by da3dsoul** [comment]
> This seems really janky man, are you trying to merge this pull request? This is absolutely not ready for prime time. Have you only tested this on your own machine? Does this literally only work in your environment?

As per the issue linked, quite a few people have used it with success. Is the current state still good? By your experience, the answer is no. This PR has been open for 2.5 months, and things change a lot very quickly in this space, so it might need a whole new round of testing since it was last called "ready".

**Comment by Urammar** [comment]
Yeah that makes a lot of sense

**Comment by Urammar** [comment]
Any joy? I'm at a loss now, sadly

**Comment by Ph0rk0z** [comment]
Well that error message is indicative of improperly installed tortoise. Probably due to windows and it not being in the right venv.

**Comment by da3dsoul** [comment]
I updated the code based on changes to silero. Coqui fucked my conda environment, so...probably don't try that atm. Due to my conda env being fucked, I'm not testing Tortoise right now. I'll probably do a full reinstall later. @Ph0rk0z if you'd like to test, be my guest.

**Comment by Ph0rk0z** [comment]
I have been using bark and I have coqui as a pip to use in audio-webui, I use the same env for it and textgen and most other "nvidia" things. Thankfully it hasn't broken anything yet. For all of these I have been setting up the environments manually so that they don't install anything I don't want. I think users want the opposite experience where the script does it all for them :P

**Comment by oobabooga** [comment]
I think that it would be really nice if the code here could be abstracted into pipelines similar to what the [multimodal](https://github.com/oobabooga/text-generation-webui/tree/main/extensions/multimodal) extension does. The resulting structure would be something similar to:

```
extensions
└── tts
    ├── pipelines
    │   ├── coqui
    │   ├── elevenlabs
    │   ├── silero
    │   ├── tortoise
    │   ├── tortoise_fast
    │   └── tortoise_mrq
    └── script.py
    └── requirements.txt
```

The one-click-installer tries to automatically install the requirements for every built-in extension, and it would fail for tortoise. By compartmentalizing it like this, we could leave only the requirements for silero and elevenlabs in `extensions/tts/requirements.txt`.

Additional TTS extensions like edge_tts in https://github.com/oobabooga/text-generation-webui/pull/3199 and bark (external) could be adapted for this framework.

The caveat is that for TTS, additional UI elements are required by each pipeline, like the API key for elevenlabs. So the framework would have to have its own `ui()` sub-functions.

I'll try to do it eventually, but if @da3dsoul wants to beat me to it that would be helpful.

**Comment by da3dsoul** [comment]
Maybe. I'll look at the multi-modal pipeline thing

**Comment by 78Alpha** [comment]
Thought I would test it out, but I haven't had much luck with llama 2 models and this working. Updated some dependencies and it says it makes a message, but all I see is it saying `Loading autoregressive model: models/tortoise\autoregressive.pth`

No error that I can see, but no output.
Updated dependencies were numba, librosa, and transformers. All to latest as of today. I turned off the unload LLM model and turned off Low VRAM to try to get it to load both, but I am still sitting at 9.1 GB / 24 GB so it seems the model never finished loading.

Added in some debug printing and it ran into `Error(s) in loading state_dict for UnifiedVoice:
        Unexpected key(s) in state_dict:`

Followed by the GPT keys.

**Comment by da3dsoul** [comment]
Good to know it doesn't work. I need to reinstall my whole workspace to fix things, as I'm currently stuck in dependency hell

**Comment by oobabooga** [comment]
I'm closing this in favor of https://github.com/oobabooga/text-generation-webui/pull/4673.

I don't want to include tortoise as XTTSv2 seems to be better overall. I don't know if the preprocessing code here applies to the new model; if so, a new PR would be welcome.

**Comment by da3dsoul** [comment]
That's fair. We can maybe re-evaluate Coqui later. Considering Coqui is what messed up my build, I'm pretty eh on it.
I can probably do some work on generifying the pre-preprocessing like I did here, maybe improving it further. We'll see if I get some time to play with it

**Pull request closed without merging**